src/hooks/useUserStats.ts

```diff
@@ -0,0 +1,131 @@
+import React from 'react';
+
+import gql from 'graphql-tag';
+import { Client } from 'urql';
+import { DocumentNode } from 'graphql';
+
+import { WPOKT_SUBGRAPH_URL } from 'constants/index';
+import { BigNumber } from 'bignumber.js';
+
+import dayjs from 'dayjs';
+
+const RETRY_EVERY = 3000;
+const ZERO = new BigNumber(0);
+const MAX_MULTIPLIER = 3;
+
+const graphqlClient = new Client({ url: WPOKT_SUBGRAPH_URL ?? '' });
+
+const USER_STATS_QUERY: DocumentNode = gql`
+  query USER_STATS($userAddress: ID!, $farmAddress: ID!) {
+    tokenGeysers(where: { id: $farmAddress }) {
+      bonusPeriodSec
+      globalSharesSec
+      createdTimestamp
+    }
+    users(where: { id: $userAddress }) {
+      id
+      earned
+      stakes {
+        amount
+        timestamp
+      }
+    }
+  }
+`;
+
+type BigNumberish = BigNumber.Value;
+
+type Stake = {
+  amount: BigNumber;
+  shares: BigNumber;
+  timestamp: number;
+};
+
+type UserStatsResponse = {
+  earned: BigNumberish;
+  stakes: [Stake];
+};
+
+type UserStatsReturnType = {
+  earned: BigNumber;
+  stakes: [Stake];
+  totalStaked: BigNumber;
+  ownershipShare: number;
+  weightedMultiplier: number;
+};
+
+export function useUserStats(userAddress: string, farmAddress: string): UserStatsReturnType {
+  const [earned, setEarned] = React.useState(ZERO);
+  const [stakes, setStakes] = React.useState<[Stake]>([{ amount: ZERO, shares: ZERO, timestamp: 0 }]);
+  const [totalStaked, setTotalStaked] = React.useState(ZERO);
+  const [ownershipShare, setOwnershipShare] = React.useState(0);
+  const [weightedMultiplier, setWeightedMultiplier] = React.useState(0);
+
+  React.useEffect(() => {
+    let cancelled = false;
+    let retryTimer: ReturnType<typeof setTimeout>;
+
+    async function fetchUserStats() {
+      try {
+        const result = await graphqlClient.query(USER_STATS_QUERY, { userAddress, farmAddress }).toPromise();
+        if (!result?.data) {
+          return;
+        }
+
+        const [{ earned, stakes }]: [UserStatsResponse] = result.data.users;
+
+        const [{ globalSharesSec, bonusPeriodSec, createdTimestamp }]: [
+          { globalSharesSec: BigNumber; bonusPeriodSec: number; createdTimestamp: number },
+        ] = result.data.tokenGeysers;
+
+        const today = dayjs();
+        const parsedEarned = new BigNumber(earned);
+
+        let totalStakeShareSecs = ZERO;
+        let parsedTotalStaked = ZERO;
+        for (let index = 0; index < stakes.length; index++) {
+          const stake: Stake = stakes[index];
+          stake.amount = new BigNumber(stake.amount);
+          parsedTotalStaked = parsedTotalStaked.plus(stake.amount);
+          totalStakeShareSecs = totalStakeShareSecs.plus(stake.amount.times(today.unix() - stake.timestamp));
```

Comment:
Hmm, this is wrong. We should not mutate React's state ourselves, but rather use the `setStakes` function given by the `useStake` hook. What I'd do is create a new stake array by looping over the existing one, updating the values, and then setting that new array as the new state.

Reply:
this is so basic, is the intern writing the app :)